ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1129,6 +1282,186 @@ private void applyDelete(MessageInfo messageInfo, RemoteReplicaInfo remoteReplic
+
+    // Example scenario: For DELETE after PUT use case in remote data center, it is possible that standby replicas get
+    // only PUT record in its replication cycle (DELETE record will come in next cycle) while leader gets both
+    // PUT and DELETE together in its replication cycle. Due to that, leader doesn't fetch
+    // the deleted blob from remote data center and the blob is never replicated from leader to standby.
+    // As a result, the PUT record in standby's missing blobs set is never emptied.
+
+    // Time out period is configurable via replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds. If
+    // replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds == -1, this safety feature is disabled.
+
+    List<RemoteReplicaInfo> remoteReplicasTimedOut = new ArrayList<>();
+    if (replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds != -1) {
+      for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+        ExchangeMetadataResponse exchangeMetadataResponse = remoteReplicaInfo.getExchangeMetadataResponse();
+        if (exchangeMetadataResponse.hasMissingStoreMessages()
+            && (time.seconds() - exchangeMetadataResponse.metadataReceivedTimeSec)
+            > replicationConfig.replicationStandbyWaitTimeoutToTriggerCrossColoFetchSeconds) {
+          remoteReplicasTimedOut.add(remoteReplicaInfo);
+        }
+      }
+    }
+    return remoteReplicasTimedOut;
+  }
+
+  /**
+   * Compare message infos of remote standby replica (whose blobs are now received from leader replicas) with message info
+   * of blobs in local store and reconcile blob properties like ttl_update, delete, undelete. If blobs for all the missing messages
+   * of the standby replica are received and updated, move the remote token of the standby forward.
+   * @param remoteReplicaInfo remote replica information
+   */
+  void processMissingKeysFromPreviousMetadataResponse(RemoteReplicaInfo remoteReplicaInfo) {
+    try {
+      ExchangeMetadataResponse exchangeMetadataResponse = remoteReplicaInfo.getExchangeMetadataResponse();
+      if (!exchangeMetadataResponse.isEmpty()) {
+
+        Set<MessageInfo> receivedStoreMessagesWithUpdatesPending =
+            exchangeMetadataResponse.getReceivedStoreMessagesWithUpdatesPending();
+        Set<MessageInfo> receivedMessagesWithUpdatesCompleted = new HashSet<>();
+
+        // 1. Go through messages whose blobs are received now (via other replicas) and compare blob metadata of
+        // remote message info with local blob in store and reconcile delete, ttl_update and undelete states
+        for (MessageInfo messageInfo : receivedStoreMessagesWithUpdatesPending) {
+          BlobId localStoreKey =
+              (BlobId) exchangeMetadataResponse.remoteKeyToLocalKeyMap.get(messageInfo.getStoreKey());
+          if (localStoreKey != null) {
+            applyUpdatesToBlobInLocalStore(messageInfo, remoteReplicaInfo, localStoreKey);
```

Comment:
Don't worry about this case, since the `applyTtlUpdate` method already catches the `Already_Updated` exception. We will not see the 'Already_Updated` and `ID_Deleted` exception being promoted here.